### DISCUSSION TITLE
How to use this pattern for project with libraries, but without executables?

### DISCUSSION THREAD

**Discussion by ArtemChandragupta** [srid/haskell-flake#369] [Q&A]
Hi. There is a project [waterfall-cad](https://github.com/joe-warren/opencascade-hs), it consists of two libraries - main library waterfall-cad (adds functional language in Haskell for 3d modelling) and it's dependency - opencascade-hs (adds Haskell bindings to opencascade-hs). The last is broken in nixpkgs, so my friend created a [fork fixing the issue](https://github.com/KovalevDima/opencascade-hs/tree/main).

I tried to build it using a pattern described here, using "hello world" placeholder executable which "depends" on opencascade-hs and waterfall-cad:

<details>

flake.nix:
```nix
{

inputs = {
  nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
  flake-parts.url = "github:hercules-ci/flake-parts";
  haskell-flake.url = "github:srid/haskell-flake";
  opencascade-hs-fork = {
    url = "github:KovalevDima/opencascade-hs";
    flake = false;
  };
};

outputs=inputs@{self,nixpkgs,flake-parts,...}:flake-parts.lib.mkFlake{inherit inputs;}{
  systems = nixpkgs.lib.systems.flakeExposed;
  imports = [ inputs.haskell-flake.flakeModule ];

  perSystem = { self', pkgs, ... }: {
    haskellProjects.default = {
      packages = {
        opencascade-hs.source = inputs.opencascade-hs-fork + /opencascade-hs;
      };
    };
    packages.default = self'.packages.example;
  };
};

}
```

example.cabal:
```cabal
cabal-version:   3.0
name:            example
version:         0.1.0.0
license:         NONE
author:          Joe
maintainer:      joe@example.com
build-type:      Simple

common warnings
    ghc-options: -Wall

executable example
    import:           warnings
    main-is:          Main.hs
    build-depends:
        base,
        opencascade-hs,
        waterfall-cad
    hs-source-dirs:   src
    default-language: Haskell2010
```

src/Main.hs:
```hs
module Main where

main :: IO ()
main = putStrLn "Hello, Haskell!"
```

</details>

It builds up, but have two problems:
1) There is useless executable
2) Haskell don't know about the libraries (modules does not exported?)
![2024-10-26-221538-window](https://github.com/user-attachments/assets/16c00f1b-a417-454a-9642-001d7c34a3c0)

**Comment by srid** [comment]
Well, you can run `ghc-pkg list` in the devShell to see what versions of libraries cabal/ghc actually will use. 

What are you doing here is right:

```nix
    haskellProjects.default = {
      packages = {
        opencascade-hs.source = inputs.opencascade-hs-fork + /opencascade-hs;
      };
    };
```

I'd guess that perhaps your devShell is stale.
